daliuge-engine/test/test_drop.py

```diff
@@ -747,8 +773,125 @@ def checkDropStates(aStatus, dStatus, eStatus, lastByte):
         checkDropStates(
             DROPStates.COMPLETED, DROPStates.COMPLETED, DROPStates.COMPLETED, b"k"
         )
+        self.assertEqual(b"ejk", droputils.allDropContents(d))
+
+    def test_objectAsNormalAndThreadedAsyncStreamingInput(self):
+        self._test_objectAsNormalAndAsyncStreamingInput(False)
+
+    def test_objectAsNormalAndMPAsyncStreamingInput(self):
+        self._test_objectAsNormalAndAsyncStreamingInput(True)
+
+    def _test_objectAsNormalAndAsyncStreamingInput(self, use_multiprocessing):
+        """
+        A test that checks that a DROP can act as normal and async streaming input
+        of different AppDROPs at the same time. We use the following graph:
+
+        A --|--> B --> D
+            |--> C --> E
+
+        Here B uses A as a streaming input, while C uses it as a normal
+        input
+        """
+
+        class QueueAsyncStream(AsyncIterator):
+            _q = multiprocessing.Queue()
+            _end = multiprocessing.Event()
+            async def __anext__(self):
+                while True:
+                    if self._q.qsize() > 0:
+                        return self._q.get()
+                    elif self._end.is_set():
+                        raise StopAsyncIteration
+                    else:
+                        await asyncio.sleep(0)
+
+            def append(self, data):
+                while not self._q.full():
+                    try:
+                        self._q.put(data, block=True, timeout=1)
+                        break
+                    except queue.Full:
+                        pass
+
+            def end(self):
+                self._end.set()
+
+        class LastCharWriterApp(AppDROP):
+            # Note: cannot share string member with test thread
+            _lastByte = multiprocessing.Value(ctypes.c_char, b" ")
+            _stream = QueueAsyncStream()
+
+            def run(self):
+                asyncio.run(self.arun())
+
+            async def arun(self):
+                outputDrop = self.outputs[0]
+                async for data in self._stream:
+                    self._lastByte.value = data[-1:]
+                    outputDrop.write(self._lastByte.value)
+
+            def dataWritten(self, uid, data):
+                self._stream.append(data)
+
+            def dropCompleted(self, uid, drop_state):
+                self._stream.end()
```

Comment:
here's another approach to stream objects using the existing streaming interface. This works with python threading and multiprocessing just has a few catches regarding memory sharing with test code.